pendulum_controller/test/test_pendulum_controller.cpp

```diff
@@ -0,0 +1,183 @@
+TEST_F(InitNodesTest, test_config) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  ASSERT_EQ(feedback_matrix, config.get_feedback_matrix());
+}
+
+// Test constructor
+TEST_F(InitNodesTest, test_constructor) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+
+  std::vector<double> state = controller.get_state();
+  std::vector<double> reference = controller.get_reference();
+  PendulumController::Config configObj = controller.get_config();
+
+  ASSERT_EQ(state[0], 0.0);
+  ASSERT_EQ(state[1], 0.0);
+  ASSERT_EQ(state[2], M_PI);
+  ASSERT_EQ(state[3], 0.0);
+
+  ASSERT_EQ(reference[0], 0.0);
+  ASSERT_EQ(reference[1], 0.0);
+  ASSERT_EQ(reference[2], M_PI);
+  ASSERT_EQ(reference[3], 0.0);
+
+  ASSERT_EQ(configObj.get_feedback_matrix(), config.get_feedback_matrix());
+}
+
+// Test update teleop
+TEST_F(InitNodesTest, test_update_teleop) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+
+  pendulum2_msgs::msg::PendulumTeleop msg_teleop;
+  msg_teleop.cart_position = M_PI / 2;
+  msg_teleop.cart_velocity = 3.0;
+
+  controller.update_teleop_data(msg_teleop);
+
+  std::vector<double> ref = controller.get_reference();
+  ASSERT_EQ(ref[0], M_PI / 2);
+  ASSERT_EQ(ref[1], 3.0);
+  ASSERT_EQ(ref[2], M_PI);
+  ASSERT_EQ(ref[3], 0.0);
+}
+
+// Test update status
+TEST_F(InitNodesTest, test_update_status) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+
+  sensor_msgs::msg::JointState msg_status;
+  msg_status.position.emplace_back(M_PI / 2);
+  msg_status.position.emplace_back(M_PI / 2);
+  msg_status.velocity.emplace_back(3.0);
+  msg_status.velocity.emplace_back(2.0);
+
+  controller.update_status_data(msg_status);
+
+  std::vector<double> state = controller.get_state();
+  ASSERT_EQ(state[0], M_PI / 2);
+  ASSERT_EQ(state[1], 3.0);
+  ASSERT_EQ(state[2], M_PI / 2);
+  ASSERT_EQ(state[3], 2.0);
+}
+
+// Test update command
+TEST_F(InitNodesTest, test_update_command) {
+  PendulumController::Config config({-10.0000, -51.5393, 356.8637, 154.4146});
+  PendulumController controller(config);
+
+  pendulum2_msgs::msg::JointCommandStamped msg_command;
+  controller.reset();
+  controller.update_command_data(msg_command);
+  ASSERT_EQ(msg_command.cmd.force, 0.0);
+
+  sensor_msgs::msg::JointState msg_status;
+  msg_status.position.emplace_back(M_PI / 2);
+  msg_status.position.emplace_back(M_PI / 2);
+  msg_status.velocity.emplace_back(3.0);
+  msg_status.velocity.emplace_back(2.0);
+  controller.update_status_data(msg_status);
+
+  pendulum2_msgs::msg::PendulumTeleop msg_teleop;
+  msg_teleop.cart_position = M_PI / 2;
+  msg_teleop.cart_velocity = 3.0;
+  controller.update_teleop_data(msg_teleop);
+
+  controller.update_command_data(msg_command);
+  ASSERT_FLOAT_EQ(msg_command.cmd.force, 251.731);
```

Comment:
I would create arrays `msg_status`, `msg_command`, and `msg_teleop` with values taken from simulation to check several iterations. For example, we could use [this](https://gitlab.com/carlossvg/inverted_pendulum_sim) to collect some simulated values.